### PR TITLE
Functionality to load any meta path data set. Further definition of the api. Solved problem with Ranking Graph

### PR DESCRIPTION
**Pull request by MkuuWaUjinga**
**_When this branch is merged, the master branch build will work again_**

### CODE REVIEW

**Comment by Baschdl** [comment]
Merge #26 first

File: api/server.py [hunk 1]
```diff
@@ -19,10 +19,9 @@
 app.config.from_object(__name__)
 # TODO: Change for deployment, e.g. use environment variable
 app.config["SECRET_KEY"] = "grgrersg346879468"
-Session(app)
-
-CORS(app, resources={r"/*": {"origins": "http://localhost:{}".format(REACT_PORT)}})
 
+CORS(app, supports_credentials=True, resources={r"/*": {"origins": "http://localhost:{}".format(REACT_PORT)}})
```
Comment (MkuuWaUjinga) [inline api/server.py]: 🤓

File: api/server.py [hunk 2]
```diff
@@ -31,26 +34,35 @@ def receive_edge_node_types():
...
 def send_next_metapaths_to_rate():
     global mock_id
-    batchsize = 5
-    paths = [{'id': i,
-              'path': ['Phenotype', 'HAS', 'Association', 'HAS', 'SNP', 'HAS', 'Phenotype'],
-              'rating': 0.5} for i in range(mock_id, mock_id + batchsize)]
-    mock_id += batchsize
-    # TODO: Call fitting method in active_learning
+    batch_size = 5
+    meta_path_loader = MetaPathLoaderDispatcher().get_loader("Rotten Tomato")
```
Comment (Baschdl) [inline api/server.py]: I think `"Rotten Tomato"` is something for the `session` object

File: api/server.py [hunk 3]
```diff
@@ -31,26 +34,35 @@ def receive_edge_node_types():
...
+    next_batch = RandomMetaPathSelector(meta_paths=meta_paths).get_next(size=batch_size)
+    paths = [{'id': id,
+              'meta_path': meta_path,
+              'rating': 0.5} for id, meta_path in zip(range(mock_id, mock_id + batch_size), next_batch)]
+    mock_id += batch_size
     return jsonify(paths)
 
+@app.route("/get-available-datasets", methods=["GET"])
+def get_available_datasets():
+    return MetaPathLoaderDispatcher().get_available_datasets()
```
Comment (Baschdl) [inline api/server.py]: Please add a endpoint description as a method comment with `""" XYZ """`

**Comment by Baschdl** [comment]
The two outdated comments aren't outdated in my opinion 😁

**Comment by feeds** [comment]
I added quite something more to this pllrequest, maybe you could review @MkuuWaUjinga

File: api/server.py [hunk 4]
```diff
@@ -26,27 +26,30 @@
...
-    session['username'] = data['username']
-    # TODO: data set and purpose is for now hardcoded, but should be set by the user.
-    session['dataset'] = 'Rotten Tomato'
-    session['purpose'] = 'none'
-    meta_path_loader = MetaPathLoaderDispatcher().get_loader(session['dataset'])
-    meta_paths = meta_path_loader.load_meta_paths()
-    session['meta_path_distributor'] = RandomMetaPathSelector(meta_paths=meta_paths)
-    session['meta_path_id'] = 1
-    session['rated_meta_paths'] = []
+    if request.method == 'POST':
```
Comment (Baschdl) [inline api/server.py]: Why you mix POST and GET here? Couldn't his be handled on two separate methods or does it really do the same stuff?

Comment (Baschdl) [inline api/server.py]: Why you mix POST and GET here? Couldn't his be handled on two separate methods or does it really do the same stuff?

File: api/server.py [hunk 5]
```diff
@@ -19,12 +21,36 @@
...
+        session['meta_path_distributor'] = RandomMetaPathSelector(meta_paths=meta_paths)
+        session['meta_path_id'] = 1
+        session['rated_meta_paths'] = []
+    return jsonify({'status': 200})
+
+
+@app.route('/logout')
+def logout():
+    # TODO better filenaming
+    filename = '{}_{}_{}.json'.format(session['dataset'],session['purpose'],session['username'])
```
Comment (Baschdl) [inline api/server.py]: Add some unique ID. Otherwise the files with the same purpose and username get overwritten.

File: api/server.py [hunk 6]
```diff
@@ -19,12 +21,36 @@
...
+    path = os.path.join(RATED_DATASETS_PATH, filename)
+    json.dump(session['rated_meta_paths'], open(path,"w", encoding="utf8"))
```
Comment (Baschdl) [inline api/server.py]: Why don't you also save 'dataset', 'purpose' and 'username'? Without this information in the json we have to parse it from the file name which is a bit ugly.

File: api/server.py [hunk 7]
```diff
@@ -48,27 +75,38 @@ def receive_edge_node_types():
...
 def send_next_metapaths_to_rate():
-    # TODO: Check if necessary information is in request object
-    global mock_id
-    batchsize = 5
-    paths = [{'id': i,
-              'path': ['Phenotype', 'HAS', 'Association', 'HAS', 'SNP', 'HAS', 'Phenotype'],
-              'rating': 0.5} for i in range(mock_id, mock_id + batchsize)]
-    mock_id += batchsize
-    # TODO: Call fitting method in active_learning
+    batch_size = 5
```
Comment (Baschdl) [inline api/server.py]: As mentioned by @GittiHab: The batch_size could be also a parameter of the method and therefore of the call of the route.

**Comment by Baschdl** [comment]
Please consider to do multiple PRs next time. I think the fixing changes for the failing master could have been merged way earlier.

File: active_learning/meta_path_selector.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    def get_next(self, size=1) -> List[MetaPath]:
+        """
+        :return: requested number of next meta-paths to be shown.
+        """
+        idx = np.random.choice(range(len(self.meta_paths)), replace=False, p=self._prob_choose_meta_path(), size=size)
+        self.visited[idx] = self.VISITED
+        return self.meta_paths[idx]
+
+    def get_all(self) -> List[MetaPath]:
```
Comment (GittiHab) [inline active_learning/meta_path_selector.py]: Comments...

File: active_learning/meta_path_selector.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        return self.meta_paths
+
+    def get_all_unrated(self) -> List[MetaPath]:
```
Comment (GittiHab) [inline active_learning/meta_path_selector.py]: Comments...

File: api/server.py [hunk 10]
```diff
@@ -19,12 +21,36 @@
...
-
+CORS(app, supports_credentials=True, resources={r"/*": {"origins": "http://localhost:{}".format(REACT_PORT)}})
 
 def run(port, hostname, debug_mode):
-    app.run(host=hostname, port=port, debug=debug_mode)
-
+    app.run(host=hostname, port=port, debug=debug_mode, threaded=True)
+
+@app.route('/login', methods=["POST", "GET"])
+def login():
```
Comment (GittiHab) [inline api/server.py]: Comment

File: api/server.py [hunk 11]
```diff
@@ -19,12 +21,36 @@
...
+        meta_path_loader = MetaPathLoaderDispatcher().get_loader("Rotten Tomato")
+        meta_paths = meta_path_loader.load_meta_paths()
+        session['meta_path_distributor'] = RandomMetaPathSelector(meta_paths=meta_paths)
+        session['meta_path_id'] = 1
+        session['rated_meta_paths'] = []
+    return jsonify({'status': 200})
+
+
+@app.route('/logout')
+def logout():
```
Comment (GittiHab) [inline api/server.py]: Comment

File: util/meta_path_loader.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
+from abc import ABC, abstractmethod
+from .datastructures import MetaPath
+from typing import List, Dict
+import pandas as pd
+from .config import ROTTEN_TOMATO_PATH
+import os
+
+class AbstractMetaPathLoader(ABC):
```
Comment (GittiHab) [inline util/meta_path_loader.py]: Comments...

File: util/meta_path_loader.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        nodes = [item[0] for item in eval(node_types)]
+        edges = eval(edge_types)
+        return MetaPath(nodes, edges)
+
+
+class RottenTomatoMetaPathLoader(AbstractMetaPathLoader):
+
+    dataset_filename = 'rotten-(when_harry_met_sally-sleepless_in_seattle)-mp1-5.csv'
+
+    def load_meta_paths(self) -> List[MetaPath]:
```
Comment (GittiHab) [inline util/meta_path_loader.py]: What does this method do? The code is really hard to read...

File: active_learning/active_learner.py [hunk 14]
```diff
@@ -1,85 +1,15 @@
...
-from typing import List
-import numpy as np
+from .hypothesis import Hypothesis
+from .meta_path_selector import RandomMetaPathSelector
+from .oracles import MockOracle
+from util.datastructures import MetaPathRatingGraph
+
 
 # algorithm types
 NO_USER_FEEDBACK = 'no_user_feedback'
```
[the remark this answers is not in the document]
Comment (Baschdl) [reply active_learning/active_learner.py]: But for logging and errors this is more descriptive

File: active_learning/meta_path_selector.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
+import numpy as np
+from typing import List
+from util.datastructures import MetaPath
+from .hypothesis import Hypothesis
+from abc import ABC, abstractmethod
+
+class AbstractMetaPathSelector(ABC):
+    """
```
Comment (Baschdl) [inline active_learning/meta_path_selector.py]: Method comments have to be in the method

Comment (MkuuWaUjinga) [reply active_learning/meta_path_selector.py]: this comment describes what the class does

Comment (Baschdl) [reply active_learning/meta_path_selector.py]: Oh sorry

**Pull request merged**